### ISSUE TITLE
[Feature request] Add function to do timezone aware DateTime calculation from weekday

### ISSUE THREAD

**Issue by baybal** [SpiritIT/timezonecomplete#42]
What I want to get is an easy way to get first DateTime of the next week in a timezone aware manner.

So, I want something like getDateTimeAtWeekday(0), or getDateTimeAtWeekdayFromTimestamp(0,anotherDateTime)

**Comment by rogierschouten** [comment]
I'm not sure I get exactly what you mean, can you give an example with the inputs and outputs? 

Would @dt.startOfDay().addLocal(tc.days(7))@ do the trick?    Of course you can add more or less than 7 depending on @dt.weekDay()@

**Comment by baybal** [comment]
Ok, the point is to determine if a given timestamp is in between Monday at, say, 7:00, and Monday 18:00, or at any other weekday.

To determine the timestamp of Monday 7:00 and Monday 18:00 in a timezone aware manner and make DateTime objects for them for comparison with greaterThen() you have to do quite a lot of calculations.

**Comment by baybal** [comment]
Something like
```
var monday0700DateTime = DateTime.fromWeekDayOf(timestampDateTimeToTest,'monday','07:00')
var monday1800DateTime = DateTime.fromWeekDayOf(timestampDateTimeToTest,'monday','18:00')

if(
    timestampDateTimeToTest.greaterEqual(monday0700DateTime)
    &&
    timestampDateTimeToTest.lessThan(monday1800DateTime)
) return true;
```

**Comment by rogierschouten** [comment]
I'm wondering whether the question 'is this time between a monday 7 and monday 18' is very common; what more generic functions would help someone trying to build this?

I'm thinking this is something for the Period class maybe.   You can already make a weekly period for a time zone starting on a known Monday at 7:00 and request the next timestamp from a given one, so the functionality is already there really. Would that help you?

i.e.  

```typescript
const reference = new tc.DateTime("some known monday in my time zone at 7:00")
const p = new tc.Period(reference, tc.days(7), tc.PeriodDst.RegularLocalTime);

const monday7 = p.findPrev(timstampDateTimeToTest)
```

Another thing that can help you is functions like the `weekdayOnOrBefore()` and `weekdayOnOrAfter`

I understand that even with these primitives it's still quite verbose to do what you want.  So what multi-purpose blocks can we come up with that would help you solve your problem and also help others solve more of theirs?

**Comment by rogierschouten** [comment]
By the way, I think you forgot some corner cases in your example code:

```typescript
var monday0700DateTime = DateTime.fromWeekDayOf(timestampDateTimeToTest,'monday','07:00')
var monday1800DateTime = DateTime.fromWeekDayOf(timestampDateTimeToTest,'monday','18:00')

if(
    timestampDateTimeToTest.greaterEqual(monday0700DateTime)
    &&
    timestampDateTimeToTest.lessThan(monday1800DateTime)
) return true;

```

if the timestamp is indeed between monday 7-18, monday0700DateTime would become the monday AFTER your timestamp and your if statement would return false

**Comment by baybal** [comment]
> I'm wondering whether the question 'is this time between a monday 7 and monday 18' is very common; what more generic functions would help someone trying to build this?

I think they are, but there are no way to replicate their functions with existing methods without rewriting a big chunk of internal logic, and that's my point.

The trickiest part is making it in a DST/TZ aware manner, down to leap seconds and such details. That's critically important in my use case.

You already have firstweekdayofmonth that does something similar.

> if the timestamp is indeed between monday 7-18, monday0700DateTime would become the monday AFTER your timestamp and your if statement would return false

I was assuming that timestampDateTimeToTest is already set prior to when the code is run. I hope I understood you right there.

**Comment by rogierschouten** [comment]
If you need leap seconds, beware:  unix timestamps do NOT carry leap seconds (and hence they are about 45 seconds off by now); and therefore also timezonecomplete cannot use them - it would cause confusion all over.  AFAIK there aren't many implementations that handle them.

**Comment by rogierschouten** [comment]
If you still want a function implemented, can you give me the exact parameter descriptions, pre and post conditions?   Just so I'm sure we're on the same page.

**Comment by baybal** [comment]
Ok, will do this tomorrow

**Comment by baybal** [comment]
1. DateTime.fromWeekDayOf(weekOf DateTime instance, WeekDay string, time string, optional weekOffset signed integer)

Produces DateTime on a given weekday of the week of "weekOf" DateTime instance. "weekOffset" allows to do the same for a week following or preceding the week of "weekOf"

2. DateTime.fromWeekDayOfUTC(DateTime instance,WeekDay string, time string, optional weekOffset signed integer)

Same, but for UTC values of both DateTimes

**Comment by rogierschouten** [comment]
Thank you.  And when is the 'weekOf' datetime considered to be 'in' a week?  Different countries have different week numbering schemes and also they have a different starting point for the day of week (e.g. some countries start weeks on Sundays and some on Mondays). 

Would it be an option to give you the first timestamp that meets the weekday & time arguments greater/equal to the 'weekOf' date?

I think we should also guarantee that for a time that exists multiple times (e.g. 2AM on a backward DST change in the Netherlands occurs twice in a day), you get the FIRST one that is greater than weekOf

Another question: so far, timezonecomplete uses three or four numbers to indicate a time, would that be an option instead of the time string?  If not, what should the allowed formats for the string be?  HH:mm?   HH:mm:ss.SSS?

**Comment by baybal** [comment]
> Thank you. And when is the 'weekOf' datetime considered to be 'in' a week? Different countries have different week numbering schemes and also they have a different starting point for the day of week (e.g. some countries start weeks on Sundays and some on Mondays).

Can we do European week, and Javascript's numbering convention first?

> Would it be an option to give you the first timestamp that meets the weekday & time arguments greater/equal to the 'weekOf' date?

Yes, that should do it.

```
var monday0700DateTime = DateTime
    .greaterThanOrEqualWeekDayOf(timestampDateTimeToTest,'monday','07:00')

var monday1800DateTime = DateTime
    .lessThanWeekDayOf(timestampDateTimeToTest,'monday','18:00')

if(
    timestampDateTimeToTest.greaterEqual(monday0700DateTime)
    &&
    timestampDateTimeToTest.lessThan(monday1800DateTime)
) return true;
```

> I think we should also guarantee that for a time that exists multiple times (e.g. 2AM on a backward DST change in the Netherlands occurs twice in a day), you get the FIRST one that is greater than weekOf

Yes, maybe add an option or a separate function that doesn't observe DST for people who don't need it.

> Another question: so far, timezonecomplete uses three or four numbers to indicate a time, would that be an option instead of the time string? If not, what should the allowed formats for the string be? HH:mm? HH:mm:ss.SSS?

Definitely. You are using LDML convention, right?

**Comment by rogierschouten** [comment]
Can you check if this fits your needs?

```typescript
import * as tc from "timezonecomplete";


/**
 * Options for timestampOnWeekTimeGreaterThanOrEqualTo() and timestampOnWeekTimeLessThan()
 */
export interface WeekTimeOpts {
	/**
	 * Timestamp to use as a basis
	 */
	reference: tc.DateTime;
	/**
	 * Desired day of week
	 */
	weekday: tc.WeekDay;
	/**
	 * Desired time (hours 0-23)
	 */
	hour: number;
	/**
	 * Desired time (minutes 0-59)
	 */
	minute: number;
}

/**
 * Returns the first timestamp >= `opts.reference` that matches the given weekday and time. Uses the time zone and DST settings
 * of the given reference time.
 * @param opts
 * @throws timezonecomplete.Argument.Hour if opts.hour out of range
 * @throws timezonecomplete.Argument.Minute if opts.minute out of range
 * @throws timezonecomplete.Argument.Weekday if opts.weekday out of range
 */
export function timestampOnWeekTimeGreaterThanOrEqualTo(opts: WeekTimeOpts): DateTime {
	assert(opts.hour >= 0 && opts.hour < 24, "Argument.Hour", "opts.hour should be within [0..23]");
	assert(opts.minute >= 0 && opts.minute < 60, "Argument.Minute", "opts.minute should be within [0..59]");
	assert(opts.weekday >= 0 && opts.weekday < 7, "Argument.Weekday", "opts.weekday should be within [0..6]");
	let midnight = opts.reference.startOfDay();
	while (midnight.weekDay() !== opts.weekday) {
		midnight = midnight.addLocal(tc.days(1));
	}
	const dt = new tc.DateTime(midnight.year(), midnight.month(), midnight.day(), opts.hour, opts.minute, 0, 0, opts.reference.zone());
	if (dt < opts.reference) {
		// we've started out on the correct weekday and the reference timestamp was greater than the given time, need to skip a week
		return dt.addLocal(tc.days(7));
	}
	return dt;
}

/**
 * Returns the first timestamp < `opts.reference` that matches the given weekday and time. Uses the time zone and DST settings
 * of the given reference time.
 * @param opts
 * @throws timezonecomplete.Argument.Hour if opts.hour out of range
 * @throws timezonecomplete.Argument.Minute if opts.minute out of range
 * @throws timezonecomplete.Argument.Weekday if opts.weekday out of range
 */
export function timestampOnWeekTimeLessThan(opts: WeekTimeOpts): DateTime {
	assert(opts.hour >= 0 && opts.hour < 24, "Argument.Hour", "opts.hour should be within [0..23]");
	assert(opts.minute >= 0 && opts.minute < 60, "Argument.Minute", "opts.minute should be within [0..59]");
	assert(opts.weekday >= 0 && opts.weekday < 7, "Argument.Weekday", "opts.weekday should be within [0..6]");
	let midnight = opts.reference.startOfDay().addLocal(days(1));
	while (midnight.weekDay() !== opts.weekday) {
		midnight = midnight.subLocal(tc.days(1));
	}
	const dt = new tc.DateTime(midnight.year(), midnight.month(), midnight.day(), opts.hour, opts.minute, 0, 0, opts.reference.zone());
	if (dt >= opts.reference) {
		// we've started out on the correct weekday and the reference timestamp was less than the given time, need to skip a week
		return dt.subLocal(tc.days(7));
	}
	return dt;
}

```

It's not as performant as it could be, but it fits these unit tests:

```typescript

describe("timestampOnWeekTimeGreaterThanOrEqualTo()", (): void => {
	it("should return the given reference time if it matches", (): void => {
		const result = timestampOnWeekTimeGreaterThanOrEqualTo({
			reference: new DateTime("2019-12-04T10:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 10,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-12-04T10:00:00.000 Europe/Amsterdam");
	});
	it("should return the given reference time if it matches (midnight)", (): void => {
		const result = timestampOnWeekTimeGreaterThanOrEqualTo({
			reference: new DateTime("2019-12-04T00:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 0,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-12-04T00:00:00.000 Europe/Amsterdam");
	});
	it("should return the next time on the same day", (): void => {
		const result = timestampOnWeekTimeGreaterThanOrEqualTo({
			reference: new DateTime("2019-12-04T10:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 11,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-12-04T11:00:00.000 Europe/Amsterdam");
	});
	it("should return the next week", (): void => {
		const result = timestampOnWeekTimeGreaterThanOrEqualTo({
			reference: new DateTime("2019-12-04T12:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 11,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-12-11T11:00:00.000 Europe/Amsterdam");
	});
});

describe("timestampOnWeekTimeLessThan()", (): void => {
	it("should NOT return the given reference time if it matches", (): void => {
		const result = timestampOnWeekTimeLessThan({
			reference: new DateTime("2019-12-04T10:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 10,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-11-27T10:00:00.000 Europe/Amsterdam");
	});
	it("should NOT return the given reference time if it matches (midnight)", (): void => {
		const result = timestampOnWeekTimeLessThan({
			reference: new DateTime("2019-12-04T00:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 0,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-11-27T00:00:00.000 Europe/Amsterdam");
	});
	it("should return the previous time on the same day", (): void => {
		const result = timestampOnWeekTimeLessThan({
			reference: new DateTime("2019-12-04T10:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 9,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-12-04T09:00:00.000 Europe/Amsterdam");
	});
	it("should return the last week", (): void => {
		const result = timestampOnWeekTimeLessThan({
			reference: new DateTime("2019-12-04T12:00:00.000 Europe/Amsterdam"),
			weekday: WeekDay.Wednesday,
			hour: 13,
			minute: 0
		});
		expect(result.toString()).to.equal("2019-11-27T13:00:00.000 Europe/Amsterdam");
	});
});

```


Please let me know.

**Comment by baybal** [comment]
Looks exactly what I need. Maybe just add an optional seconds to WeekTimeOpts and it should be it.

**Comment by rogierschouten** [comment]
ok, will add this!

**Comment by rogierschouten** [comment]
published as v5.11.0
